src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs

```diff
@@ -0,0 +1,233 @@
+	{
+		// we want to ensure several things:
+		// 1. All enum values are marked with a Field attribute
+		// 2. All Field attributes have a symbol name
+		// 3. All symbol names have to be unique
+		// 4. If the Field attribute is not from a known apple library, the library name is set
+		// 5. If the Field attribute is from a known apple library the lib should be null
+		// 6. The user used the correct attribute flag.
+
+		// based on the platform decide if we are dealing with a known apple framework, we want all, not just the
+		// ones that are part of the simulator
+		if (platformName == PlatformName.None) {
+			// we could not identify the platform, we have a bug or we are not on an apple target, do nothing
+			return [];
+		}
+
+		var appleFrameworks = Frameworks.GetFrameworks (platformName.ToApplePlatform (), false);
+		if (appleFrameworks is null) {
+			// we could not get the frameworks, we have a bug
+			return [];
+		}
+
+		var isAppleFramework = appleFrameworks.Find (symbol.ContainingNamespace.Name) is not null;
+
+		// bucket with all the diagnostics we have found
+		var bucket = ImmutableArray.CreateBuilder<Diagnostic> ();
+
+		var members = symbol.GetMembers ().OfType<IFieldSymbol> ().ToArray ();
+		// we do not allow backing fields to be duplicated, keep track of those added so far to raise
+		// a diagnostic if we find a duplicate
+		var backingFields = new HashSet<string> ();
+		foreach (var fieldSymbol in members) {
+			var attributes = fieldSymbol.GetAttributeData ();
+			if (attributes.Count == 0) {
+				//  All enum values are marked with a Field attribute, therefore add a diagnostic
+				bucket.Add (Diagnostic.Create (
+					RBI0002, // Smart enum values must be tagged with an Field<EnumValue> attribute.
+					fieldSymbol.Locations.First (),
+					fieldSymbol.ToDisplayString ()));
+				continue;
+			}
+
+			// Get all the FieldAttribute, parse it and add the data to the result
+			if (attributes.TryGetValue (AttributesNames.EnumFieldAttribute, out var fieldAttrDataList)
+				&& fieldAttrDataList.Count == 1) {
+				var fieldAttrData = fieldAttrDataList [0];
+				var fieldSyntax = fieldAttrData.ApplicationSyntaxReference?.GetSyntax ();
+				if (fieldSyntax is null) {
+					continue;
+				}
+
+				if (FieldData<EnumValue>.TryParse (fieldAttrData, out var fieldData, out var errorTuple)) {
+					// only provide diagnostics if we managed to parse the FieldData, else we have a bug in the
+					// analyzer
+					if (!backingFields.Add (fieldData.Value.SymbolName)) {
+						// All symbol names have to be unique
+						bucket.Add (Diagnostic.Create (
+							RBI0003, // Smart enum backing field cannot appear more than once.
+							fieldSyntax.GetLocation (),
+							fieldSymbol.ToDisplayString (), fieldData.Value.SymbolName));
+					}
+
+					if (!isAppleFramework) {
+						// If the Field attribute is not from a known apple library, the library name is set
+						if (string.IsNullOrWhiteSpace (fieldData.Value.LibraryName)) {
```

Comment:
This test belies your error message. You probably want to change your error message to be "null or empty" or break out your cases into:
empty, obviously incorrect path, likely correct path. And if you go this route, I'd break it out into a separate function:
```csharp
bool TryGetErrorForLibrary (bool isAppleFrameWork, string? libraryName, [NotNullWhen(true)] out Diagnostic? error)
{
     // implementation here
}
// then your call site here is:
if (TryGetErrorForLibrary (isAppleFrameWork, fieldData.Value.LibraryName, out var libraryError))
    bucket.Add (libraryError);
```